cpp/include/cudf/filling.hpp

```diff
@@ -223,13 +224,36 @@ std::unique_ptr<column> sequence(
  * @param months Months to increment
  * @param mr Device memory resource used to allocate the returned column's device memory
  *
- * @returns Timestamps column with sequences of months.
+ * @return Timestamps column with sequences of months.
  */
 std::unique_ptr<cudf::column> calendrical_month_sequence(
   size_type size,
   scalar const& init,
   size_type months,
   rmm::mr::device_memory_resource* mr = rmm::mr::get_current_device_resource());
 
+/**
+ * @brief Generate a column containing complement of the indices given from the input map columns.
+ *
+ * For a number of given columns containing indices in the range of [0, @p size) that represent some
+ * gather/scatter maps, generate an output column containing the indices in the same range which do
+ * not appear in any of the given maps.
+ *
+ * Duplicates indices have no affect on the outcome. Invalid indices (i.e., the indices that are
+ * outside of the given range [0, @p size)) are ignored during generating the output.
+ *
+ * @throws cudf::logic_error if any of the given @p maps column is not of integer types.
+ * @throws cudf::logic_error if @p size is < 0.
+ *
+ * @param maps The columns containing input indices.
+ * @param size Size that defines the range of indices in both input and output.
+ * @param mr Device memory resource used to allocate the returned column's device memory.
+ * @return The result column containing the generated complement indices.
+ */
+std::unique_ptr<cudf::column> complement(
```

Comment:
Gather maps that are used in joins are not columns.  This would need to operate on something like `rmm::device_uvector<cudf::size_type>` instead of columns.  See the types returned by the join APIs and `get_left_join_indices_complement` for an example of where libcudf is already computing this today.

Reply:
Yes, I was worried about that. That's why I made the second version that returns `device_uvector`. The public API can then construct a column from it.

Reply:
Oh I didn't look closely at the other one since it's a detail API.  Is our application supposed to call detail APIs for this?

Reply:
Our use case is just for join maps, i.e.: `device_uvector`, and we would need to expose this in the cudf Java bindings, so effectively it would be a public API, at least from the Java bindings perspective.

Reply:
It should be a `device_span` for the input instead of a `column_view`. But why are we accepting multiple inputs?

Reply:
Because we may have multiple input maps, like in the original issue description that we generate multiple gather maps separately from the partitioned data. Concatenating these maps may have the output size exceeds INT_MAX.

Reply:
So as per suggestions:
 * Input type is `device_span`
 * Output type is `device_uvector`

Reply:
> Concatenating these maps may have the output size exceeds INT_MAX.

That's not a problem for `device_span` since the `size_type` is `size_t`.

Reply:
Alright, if so, the input gather maps should be generated as `device_uvector` from somewhere else, and we need to concatenate them together into one `device_uvector` before passing a `device_span` referring to it to the API.

I'm still not clear and curious now is whether it is trivial to concatenate multiple `device_uvector`s into one `device_uvector`? The first solution that came up from my mind is to compute the total size, then create a new `device_uvector` and copy data of the input maps into it.

Even after we concatenated all the input maps into one map, one more thing we need to consider is the memory and performance overhead:
 * If we pass multiple input maps directly into the new API, we will have to launch multiple scatter kernels, each launch for one map
 * If we concatenate multiple maps into one map then pass it in, we will have only one kernel launch, but we have to pay for the overhead (and memory) for such concatenation.

@jrhemstad What is your comment for which solution above to use?

Reply:
TL;DR: I'm happy to start with an API that only takes a single gather map input.

For my use-case, which is generating the gather map complement of rows in the right table that were selected at least once in a series of join operations, the application will be generating a series of gather maps as it streams through the batches of the left table, joining each batch against the right table.  To keep the memory manageable when the left table is very, very large, I'll want to periodically concatenate and de-duplicate the gather maps before I call this complement function. Since I have to concatenate and de-duplicate anyway before calling this, I'm OK if we start with a simpler approach where the gather map complement function takes a single gather map input.

However separately that means I need a way to concatenate and de-duplicate gather map entries.  Concatenate is easy, that's just cudaMemcpy.  The deduplicate is a bit trickier but I think still manageable.  I can "window" the gather maps with column views and use column deduplicate (possibly doing multiple iterations, concatenating the intermediate results).  That leaves a corner case where the right table size is near or at the cudf row limit and most rows were selected in the join operations.  As I deduplicate and merge the batches, I could end up with two gather maps, each close to the cudf row limit in size, and at that point the column view "window" isn't big enough to effectively deduplicate directly.  If I'm stuck deduplicating with only column views, I'd have to implement a more complicated algorithm like a merge sort on all the gather map chunks.   An alternative in that rare case is I can just take the memory hit, concatenate them as they are, and call the gather map complement function on the large result.

Reply:
FYI: Since you proposed to iteratively concatenate and drop duplicates, I have filed a new issue asking for a more efficient version of `drop_duplicates` (https://github.com/rapidsai/cudf/issues/9886).